dockertools: call TearDownPod when GC-ing infra pods

dcbw opened the pull request:
The docker runtime doesn't tear down networking when GC-ing pods.
rkt already does so make docker do it too. To ensure this happens,
infra pods are now always GC-ed rather than gating them by
containersToKeep.

This prevents IPAM from leaking when the pod gets killed for
some reason outside kubelet (like docker restart) or when pods
are killed while kubelet isn't running.

Fixes: https://github.com/kubernetes/kubernetes/issues/14940
Related: https://github.com/kubernetes/kubernetes/pull/35572

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37036)
<!-- Reviewable:end -->

dcbw committed:
dockertools: clean up networking when garbage-collecting pods

The docker runtime doesn't tear down networking when GC-ing pods.
rkt already does so make docker do it too. To ensure this happens,
networking is always torn down for the container even if the
container itself is not deleted.

This prevents IPAM from leaking when the pod gets killed for
some reason outside kubelet (like docker restart) or when pods
are killed while kubelet isn't running.

Fixes: https://github.com/kubernetes/kubernetes/issues/14940
Related: https://github.com/kubernetes/kubernetes/pull/35572

dcbw commented:
@kubernetes/sig-network @bprashanth @thockin @caseydavenport

bprashanth commented:
I'll need to page this in @kubernetes/sig-node might be able to review this right away. 
Fyi we fixed the ipam leak for kubenet with a hack for the point release: https://github.com/kubernetes/kubernetes/issues/34278#issuecomment-256181848. I think we discussed on that issue that teardown is called, but we can't find the ip of a dead container to look up its netns.

dcbw commented:
> I think we discussed on that issue that teardown is called, but we can't find the ip of a dead container to look up its netns.

@bprashanth in the case this PR fixes, teardown is *not* called, because this is container GC and the docker runtime doesn't do any network anything on GC containers...

dcbw commented:
@k8s-bot verify test this issue https://github.com/kubernetes/kubernetes/issues/36843

dcbw commented:
Jenkins verify is failing on update-bazel.sh that doesn't appear to have anything to do with this PR...

mikedanese commented:
@dcbw can you please run hack/update-hazel.sh?

dcbw commented:
@mikedanese 

`package github.com/mikedanese/gazel: cannot download, $GOPATH not set. For more details see: go help gopath`

So I add "kube::golang::setup_env" to hack/update-bazel.sh and get a ton of:

```
extract err: cannot find package "k8s.io/gengo/types" in any of:
	/usr/lib/golang/src/k8s.io/gengo/types (from $GOROOT)
	/home/dcbw/Development/containers/kubernetes/_output/local/go/src/k8s.io/gengo/types (from $GOPATH)
```

whereupon I have tons of changes that are completely unrelated to anything in this PR.  Shouldn't somebody else run update-bazel.sh and submit that as a PR, and then I rebase on top?

```
diff --git a/cmd/genman/BUILD b/cmd/genman/BUILD
index 0ff8bde..d149f5c 100644
--- a/cmd/genman/BUILD
+++ b/cmd/genman/BUILD
@@ -12,7 +12,10 @@ load(
 
 go_binary(
     name = "genman",
-    srcs = ["gen_kube_man.go"],
+    srcs = [
+        "gen_kube_man.go",
+        "gen_kubectl_man.go",
+    ],
     tags = ["automanaged"],
     deps = [
         "//cmd/genutils:go_default_library",
@@ -23,8 +26,5 @@ go_binary(
         "//pkg/kubectl/cmd:go_default_library",
         "//pkg/kubectl/cmd/util:go_default_library",
         "//plugin/cmd/kube-scheduler/app:go_default_library",
-        "//vendor:github.com/cpuguy83/go-md2man/md2man",
-        "//vendor:github.com/spf13/cobra",
-        "//vendor:github.com/spf13/pflag",
     ],
 )
diff --git a/cmd/genswaggertypedocs/BUILD b/cmd/genswaggertypedocs/BUILD
index 21347db..c266af9 100644
--- a/cmd/genswaggertypedocs/BUILD
+++ b/cmd/genswaggertypedocs/BUILD
@@ -14,9 +14,5 @@ go_binary(
     name = "genswaggertypedocs",
     srcs = ["swagger_type_docs.go"],
     tags = ["automanaged"],
-    deps = [
-        "//pkg/runtime:go_default_library",
-        "//vendor:github.com/golang/glog",
-        "//vendor:github.com/spf13/pflag",
-    ],
+    deps = ["//pkg/runtime:go_default_library"],
 )
<etc>
```

mikedanese commented:
@dcbw hmm, it looks like you have a somewhat non standard gopath setup. gazel expects kubernetes source to be checked out in $GOPATH/src/k8s.io/kubernetes. In #36843 we are adding a hard check for this. See:

https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md#clone-your-fork

In the meantime you can apply this diff to fix ci:

```diff
diff --git a/pkg/kubelet/BUILD b/pkg/kubelet/BUILD
index eaade63..73d6ebe 100644
--- a/pkg/kubelet/BUILD
+++ b/pkg/kubelet/BUILD
@@ -65,6 +65,7 @@ go_library(
         "//pkg/kubelet/eviction:go_default_library",
         "//pkg/kubelet/images:go_default_library",
         "//pkg/kubelet/kuberuntime:go_default_library",
+        "//pkg/kubelet/leaky:go_default_library",
         "//pkg/kubelet/lifecycle:go_default_library",
         "//pkg/kubelet/metrics:go_default_library",
```

dcbw commented:
> @dcbw hmm, it looks like you have a somewhat non standard gopath setup. gazel expects kubernetes source to be checked out in $GOPATH/src/k8s.io/kubernetes. In #36843 we are adding a hard check for this. See:

Ah, thanks!  So yeah, it was something in the patch.  Is the BUILD stuff new?  I guess I've either been lucky before that all imports have already been in BUILD.

RE the gopath, I don't have a particularly non-standard setup for Linux developers that I know of.  Literally 'git clone kubernetes', cd kubernetes,then make.  That has always worked, and appears supported since the purpose of "kube::golang::setup_env" and hack/lib/golang.sh is to set up the Go environment without having GOPATH set. Is that not the case?

dcbw commented:
@k8s-bot cvm gke e2e test this issue https://github.com/kubernetes/kubernetes/issues/35913

yujuhong commented:
@dcbw from reading https://github.com/kubernetes/kubernetes/issues/14940#issuecomment-253033032, since kubelet cannot recover the netns information when docker container terminated, this fix doesn't always work? Another question is that will the cni plugin returns an error if 1) netns is empty, 2) if teardown is called multiple times?

dcbw commented:
> @dcbw from reading #14940 (comment), since kubelet cannot recover the netns information when docker container terminated, this fix doesn't always work? Another question is that will the cni plugin returns an error if 1) netns is empty, 2) if teardown is called multiple times?

At least for Docker, if the container is "dead" or "exited" the netns doesn't exist anymore, so any interfaces it had will have already been torn down.  But IPAM may not have been torn down, and that still needs to happen which this PR allows the CNI plugin to do.

CNI does not require a CNI_NETNS for CNI_DEL actions, and its up to the plugin to do best-effort cleanup when it gets the CNI_DEL call.  That should probably include any IPAM to ensure that even if the netns and container are gone, the IPAM gets released.

In any case, this patch doesn't regress the DEL behavior from what it is now.  If the CNI plugin requires a netns, it'll still return an error just like it does today.  And teardown is already called multiple times due to kubelet races :)  So that's not a change in behavior either.

yujuhong commented on pkg/kubelet/pod_container_deletor.go:
```diff
@@ -80,20 +81,27 @@ func getContainersToDeleteInPod(filterContainerId string, podStatus *kubecontain
 
 	if len(candidates) <= containersToKeep {
-		return containerStatusbyCreatedList{}
+		// Always destroy infra containers to ensure networking gets torn down
```
kubelet still relies on one dead container as a checkpoint. I don't think we should change that. 

Overall, I'd prefer not making change to this file (`pod_container_deletor.go`). The podContainerDeletor simply does a more aggressive cleanup, and the regular container GC will kick in later. 

It's also a plus to not pollute more files with leaky.PodInfraContainerName

feiskyer replied:
There are actually two different moments which may trigger network teardown:

(1) when infra container is down
(2) when infra container is removed

I think codes in `removeContainer()` handles (2) and here is a fix for (1).

To reduce IPAM leaking asap, I suggest we only trigger network tear down at infra container stopped, e.g. perform infra container network teardown after getting evictableContainers (and before removing any containers).

feiskyer replied:
There are actually two different moments which may trigger network teardown:

(1) when infra container is down
(2) when infra container is removed

I think codes in `removeContainer()` handles (2) and here is a fix for (1).

To reduce IPAM leaking asap, I suggest we only trigger network tear down at infra container stopped, e.g. perform infra container network teardown after getting evictableContainers (and before removing any containers).

yujuhong replied:
@dcbw I'd like to try making the change small and localized since it's so close to the release (unless this is not targeted for 1.5).

Change in this file is very docker-specific even though it's used by all implementations (rkt, cri, etc). On top of that, kubenet has its own fix that handles the case where the infra containers get removed underneath kubelet. By leaving everything to the container GC, you have already fixed the leaks in most situation (e.g., docker restart). Would that be sufficient?

yujuhong commented on pkg/kubelet/dockertools/container_gc.go:
```diff
@@ -245,35 +261,47 @@ func (cgc *containerGC) GarbageCollect(gcPolicy kubecontainer.ContainerGCPolicy,
-	glog.V(4).Infof("Removing container %q name %q", id, containerName)
-	err := cgc.client.RemoveContainer(id, dockertypes.ContainerRemoveOptions{RemoveVolumes: true})
+func (cgc *containerGC) removeContainer(containerInfo containerGCInfo) {
+	if !containerInfo.isHostNetwork && containerInfo.containerName == leaky.PodInfraContainerName {
```
s/leaky.PodInfraContainerName/PodInfraContainerName

It's already defined in `dockertools/docker.go`

dcbw commented:
> On top of that, kubenet has its own fix that handles the case where the infra containers get removed underneath kubelet.

@yujuhong yeah, kubenet has that workaround, but it was a workaround for this specific problem :)  And it's only for kubenet, which doesn't help CNI plugins at all.  For OpenShift we just ported your kubenet workaround to our CNI plugin, but I thought I'd try to fix it for real too.

chenchun commented:
@dcbw This PR is useful, but I think it can't resolve all the situations. Here is an exception:

Suppose we are using [host-local ipam in cni plugin](https://github.com/containernetworking/cni/tree/master/plugins/ipam/host-local)

1. kubelet starts cni plugin to setup network
2. cni plugin process `cni-1` starts and it supposes to setup network
3. kubectl deletes the pod
4. kubelet starts cni plugin to teardown network
5. cni plugin process `cni-2` starts and it supposes to teardown network
6. cni plugin process `cni-2` exists
7. kubelet deletes infrastructure container
8. `cni-1` touches ip file and exists

This is because kubelet didn't hold a lock before starting cni plugin and guarantee that `cni-1` must exist before launching `cni-2`. I'm not sure if kubelet needs to hold a lock. IMO, but this may not be the responsibility of kubelet to hold a lock or to delete these leaky ip files. Anyone who uses cni plugin with host-local ipam will have to launch a demon process to cleanup these leaky ip files.

dcbw committed:
dockertools: use network PluginManager to synchronize pod network operations

We need to tear down networking when garbage collecting containers too,
and GC is run from a different goroutine in kubelet.  We don't want
container network operations running for the same pod concurrently.

dcbw commented:
@yujuhong @chenchun @bprashanth  I believe I'd addressed all your comments.  First, everything is now specific to the docker runtime, no more kubelet general changes. Second, we now serialize all network plugin operations in the docker runtime to specifically guard against concurrent operations from kubelet's GC goroutine and kubelet's SyncPod() goroutine.  This was always a problem before too and we've run into instances of concurrent teardown in the past.

@alexbrand I've made some changes to the Windows parts of dockertools too.  Could you verify them?  From the comments I understand that every container in the pod gets its own IP address on Windows, but that the Infra container (while started) is basically ignored.  So I attempted to ensure that all containers on Windows will get their networking cleaned up when GC happens, whiel on Linux only the Infra container needs this.

dcbw commented:
@kubernetes/rh-networking @kubernetes/sig-network may be relevant to these groups as well

yujuhong commented:
/cc @freehan

chenchun commented on pkg/kubelet/dockertools/docker_network.go:
```diff
@@ -0,0 +1,106 @@
+	networkPlugin network.NetworkPlugin
+
+	// Pod network operations lock
+	mu sync.Mutex
```
This is really a big lock for all pods and I wonder how much it may slow down launching/deleting pods. Why not make every pod holds its own lock?

dcbw replied:
@chenchun I implemented the per-pod locks for network operations and added testcases for locking correctness.

feiskyer commented on pkg/kubelet/dockertools/container_gc.go:
```diff
@@ -173,26 +193,34 @@ func (cgc *containerGC) evictableContainers(minAge time.Duration) (containersByE
 		err = cgc.client.RemoveContainer(container.id, dockertypes.ContainerRemoveOptions{RemoveVolumes: true})
 		if err != nil {
-			glog.Warningf("Failed to remove unidentified dead container %q: %v", container.name, err)
+			glog.Warningf("Failed to remove unidentified dead container %q: %v", container.containerName, err)
```
Containers in multiple pods may have same name, should change it to docker's full container name for clear.

dcbw replied:
@feiskyer fixed.

feiskyer commented on pkg/kubelet/dockertools/container_gc.go:
```diff
@@ -111,22 +116,41 @@ func (cgc *containerGC) removeOldestN(containers []containerGCInfo, toRemove int
+func newContainerGCInfo(id string, inspectResult *dockertypes.ContainerJSON, created time.Time) (containerGCInfo, error) {
+	containerName, _, err := ParseDockerName(inspectResult.Name)
+	if err != nil {
+		return containerGCInfo{id: id}, fmt.Errorf("failed to parse docker name: %v", err)
```
containerName is always "" in this case, which will cause missing container name in following logs.

dcbw replied:
fixed now

dcbw commented:
@feiskyer @yujuhong @chenchun PTAL thanks!

freehan commented on pkg/kubelet/dockertools/docker_network.go:
```diff
@@ -0,0 +1,154 @@
+	"k8s.io/kubernetes/pkg/kubelet/network"
+)
+
+type dockerNetwork struct {
```
This is essentially a network plugin wrapper which helps to synchronize the calls. I feels like it can be more generic than dockerNetwork, maybe NetworkPluginManager? The only docker specific stuff here is the docker network mode. How about this? only wrap GetPodNetworkStatus here and leave determineContainerIP in dockermanager?

dcbw replied:
@freehan yeah I thought about that.  I'll go ahead and do this.

dcbw committed:
kubelet: add network plugin manager with per-pod operation locking

The PluginManager almost duplicates the network plugin interface, but
not quite since the Init() function should be called by whatever
actually finds and creates the network plugin instance.  Only then
does it get passed off to the PluginManager.

The Manager synchronizes pod-specific network operations like setup,
teardown, and pod network status.  It passes through all other
operations so that runtimes don't have to cache the network plugin
directly, but can use the PluginManager as a wrapper.

dcbw committed:
kubelet/network: move mock network plugin to pkg/kubelet/network/testing

dcbw committed:
dockershim: use network PluginManager to synchronize pod network operations

dcbw committed:
rkt: use network PluginManager to synchronize pod network operations

dcbw commented on pkg/kubelet/dockertools/docker_manager_linux.go:
```diff
@@ -40,8 +40,13 @@ func getContainerIP(container *dockertypes.ContainerJSON) string {
+}
+
+// Only the infrastructure container needs network setup/teardown
+func containerIsNetworked(containerName string) bool {
```
@freehan isn't a host-network container sandboxed, just not for networking?

caseydavenport replied:
Or even just `isPodInfraContainer`?  I find the current naming a bit confusing.

dcbw commented:
@freehan @yujuhong @feiskyer @chenchun PTAL, thanks!

Latest is a pretty major refactor:

1) plugin manager bits moved to pkg/kubelet/network
2) the kubelet mock network plugin  moved to pkg/kubelet/testing since other testing stuff was already there; this also removes the "fake network plugin" I added in favor of the MockNetworkPlugin.
3) ported dockershim and rkt runtimes over to the plugin manager

chenchun commented on pkg/kubelet/network/plugins.go:
```diff
@@ -277,3 +278,121 @@ func GetPodIP(execer utilexec.Interface, nsenterPath, netnsPath, interfaceName s
+
+func (pm *PluginManager) GetPodNetworkStatus(podNamespace, podName string, id kubecontainer.ContainerID) (*PodNetworkStatus, error) {
+	fullPodName := kubecontainer.BuildPodFullName(podName, podNamespace)
+	pm.podLock(fullPodName).Lock()
```
I think you are wrongly using locks here. Each time you hold a big lock, you won't get your parallel. I thought this for a while, maybe a file lock is suitable here. But I'm still hesitating...

chenchun replied:
I wrote a quick prototype https://github.com/chenchun/keylock, I think it maybe a good solution. In a big chance, each pod gets it's owner lock and you don't have to worry about when to create/delete locks. In bad situations, when the `keylock.KeyGen` generates duplicate uint32s for different podNames, network operations for these pods will be synchronised which I think is acceptable.

dcbw replied:
@chenchun I could be completely off here, but I'm pretty sure we do get the parallel for different pods. Yes, there is a short period of time (for accessing the pod lock map) when we do block network operations for all pods, but that is a very short period compared to the actual network operations.  It's the setup/teardown/status that take the largest amount of time, and those should be executed in parallel between pods but synchronized for the same pod.

dcbw replied:
Also TestMultiPodParallelNetworkOps() ensures that pod operations can run in parallel by blocking the setup of one pod on another pod completing operations.

chenchun replied:
I think you are right. Sorry if I may have wasted your time here.

freehan commented:
Need to fix unit test and verification.

dcbw commented:
@freehan yeah, was working on that while you commented; had a file left-over in my source tree that implemented FakeNetworkPlugin.  Also just made bazel happy.

dcbw commented:
Ugh, what is it with "Unable to load build details from" from jenkins....

feiskyer commented:
> Ugh, what is it with "Unable to load build details from" from jenkins....

```
I1214 15:16:01.058] FAIL	k8s.io/kubernetes/pkg/kubelet/network/testing	2.142s
```

caseydavenport commented on pkg/kubelet/dockertools/container_gc.go:
```diff
@@ -245,35 +281,53 @@ func (cgc *containerGC) GarbageCollect(gcPolicy kubecontainer.ContainerGCPolicy,
-	glog.V(4).Infof("Removing container %q name %q", id, containerName)
-	err := cgc.client.RemoveContainer(id, dockertypes.ContainerRemoveOptions{RemoveVolumes: true})
+func (cgc *containerGC) netContainerCleanup(containerInfo containerGCInfo) {
+	if containerInfo.isHostNetwork || !containerIsNetworked(containerInfo.containerName) {
```
I'm assuming we need to check `!containerIsNetworked` because this is getting called in `deleteContainer` above?  Wondering if we could remove the need for this check within this method.  Shouldn't it really be an error to call `netContainerCleanup` on something that isn't a "net container"?

dcbw replied:
Good point, removed and added to deleteContainer().

caseydavenport commented on pkg/kubelet/dockertools/docker_manager_linux.go:
```diff
@@ -40,8 +40,13 @@ func getContainerIP(container *dockertypes.ContainerJSON) string {
 // Returns true if the container name matches the infrastructure's container name
-func containerProvidesPodIP(name *KubeletContainerName) bool {
-	return name.ContainerName == PodInfraContainerName
+func containerProvidesPodIP(containerName string) bool {
```
What's the difference between `containerProvidesPodIP` and `containerIsNetworked`, couldn't these just be a single function?

dcbw replied:
@caseydavenport This was for Windows.  If you look at docker_windows.go, it looks like every non-infra container on Windows could receive an IP address.  But since Kubernetes only allows one IP address per pod, its very unclear which container's IP gets used.

GetPodStatus() appears to just use the last container in the list for the pod IP for Windows.  I'll try to ask @alexbrand about this though...

alexbrand replied:
Yes. Due to limitations on the windows side (mainly the fact that container networking is not supported), we couldn't use the infrastructure container for pods on windows.

dcbw replied:
Yeah, @alexbrand says each container in a Windows pod gets its own IP.  So containerProvidesPodIP() isn't quite the same as containerIsNetworked(), since one container is suppposed to supply the pod IP, but all of them are networked.

caseydavenport replied:
d'oh, yeah, gotcha.

caseydavenport commented on pkg/kubelet/network/plugins.go:
```diff
@@ -277,3 +278,121 @@ func GetPodIP(execer utilexec.Interface, nsenterPath, netnsPath, interfaceName s
+// Lock network operations for a specific pod.  If that pod is not yet in
+// the pod map, it will be added.  The reference count for the pod will
+// be increased.
+func (pm *PluginManager) podLock(fullPodName string) *sync.Mutex {
```
I think this is more of a `lockPod(` than a `podLock(` :)

(and `unlockPod` below)

caseydavenport replied:
Nevermind - this does actually return the Mutex, so `podLock` seems appropriate.

caseydavenport commented on pkg/kubelet/network/plugins.go:
```diff
@@ -277,3 +278,121 @@ func GetPodIP(execer utilexec.Interface, nsenterPath, netnsPath, interfaceName s
+
+	lock, ok := pm.pods[fullPodName]
+	if !ok || lock.refcount == 0 {
+		glog.Warningf("Unbalanced pod lock unref for %s", fullPodName)
```
Don't we need to have an explicit check for `lock.refcount == 0` so we can remove it from `pm.pods`?

dcbw replied:
@caseydavenport I'm pretty sure this couldn't ever happen, but I've done this anyway to better recover if it did.

caseydavenport replied:
Cool, yeah.  I also figure it should never happen but if we're going to be handling it anyway...

caseydavenport commented on pkg/kubelet/network/plugins.go:
```diff
@@ -277,3 +278,121 @@ func GetPodIP(execer utilexec.Interface, nsenterPath, netnsPath, interfaceName s
+	netStatus, err := pm.plugin.GetPodNetworkStatus(podNamespace, podName, id)
+	if err != nil {
+		return nil, fmt.Errorf("NetworkPlugin %s failed on the status hook for pod %q: %v", pm.plugin.Name(), fullPodName, err)
+	} else if netStatus == nil {
```
Is this not a redundant branch? If netStatus is nil we'll return `nil, nil` below anyway I think.

caseydavenport commented on pkg/kubelet/network/plugins.go:
```diff
@@ -277,3 +278,121 @@ func GetPodIP(execer utilexec.Interface, nsenterPath, netnsPath, interfaceName s
+
+	glog.V(3).Infof("Calling network plugin %s to set up pod %q", pm.plugin.Name(), fullPodName)
+	if err := pm.plugin.SetUpPod(podNamespace, podName, id); err != nil {
+		// TODO: (random-liu) There shouldn't be "Skipping pod" in sync result message
```
Could we knock off an easy TODO while we're here? :)

dcbw replied:
Done

caseydavenport reviewed (commented):
@dcbw thanks a lot for working on this :)

I stuck some comments on.

dcbw commented:
> Ugh, what is it with "Unable to load build details from" from jenkins....
> 
> I1214 15:16:01.058] FAIL	k8s.io/kubernetes/pkg/kubelet/network/testing	2.142s

@feiskyer this failure was from a later run when Jenkins had finally started reporting logs.  But that still doesn't actually say what the failure is, unless I'm blind...  I can't reproduce it with "make check WHAT=pkg/kubelet/network/testing" either.

dcbw commented:
@feiskyer fixed the testcase, goroutines weren't waiting for all the others to get started.

freehan commented:
Thanks! LGTM. I will wait for others take a final look.

feiskyer commented:
LGTM

caseydavenport commented:
Yep, LGTM.

caseydavenport commented on pkg/kubelet/dockertools/container_gc.go:
```diff
@@ -245,35 +281,56 @@ func (cgc *containerGC) GarbageCollect(gcPolicy kubecontainer.ContainerGCPolicy,
-func (cgc *containerGC) removeContainer(id string, podNameWithNamespace string, containerName string) {
-	glog.V(4).Infof("Removing container %q name %q", id, containerName)
-	err := cgc.client.RemoveContainer(id, dockertypes.ContainerRemoveOptions{RemoveVolumes: true})
+func (cgc *containerGC) netContainerCleanup(containerInfo containerGCInfo) {
```
@dcbw actually, I have a slight concern.

At this point in execution, can we say that the pod that this net container used to belong to is not running? 
 
e.g is this sequence of events possible?
- Pod "nginx-abcdef" is running with infra container X 
- Docker is restarted, infra container goes away.
- Pod "nginx-abcdef" is recreated with a new infra container Y.
- Kubelet (with this PR) tears down infra container X.

I know that Calico has worked around this case for a long time by correlating some information with the namespace/podname (not sure if other plugins do anything similar).  This historically meant that the plugin could tell when a pod has been restarted (and clean up old IP allocations, for example).  However, if the above ordering is correct then the existing plugin will actually delete metadata about a currently running pod :(

It probably would mean changes to the Calico plugin (and any other plugins that might do something similar), but would also render older versions of the plugin incompatible with this change.

dcbw replied:
@caseydavenport we can't say that at this point.  The net container being cleaned up may well be a dead net container of a still-running pod that has already created a new net container. The sequence of events you post here is exactly one of the sequences I saw when testing this out.

Is there something I could do to mitigate this issue for plugin writers, and/or could it be documented for kube 1.5? Could you describe a bit more how the Calico plugin handles correlation between IPAM and specific container IDs versus pod name/namespace?

Also, does the existing Calico workaround handle the case where the pod does not get rescheduled on the same node, but the net container is still dead with its IP allocation intact?  In that case kubelet would never have called the plugin at all, either for setup or teardown, so the plugin would need some kind of node-external tracking service for the IPAM to know when the pod was rescheduled on another node and then clear out the allocation on the existing node somehow.

caseydavenport replied:
Yeah, that's what I thought.

I have to think a bit more about how this might be mitigated.  The only way I can think of right now is if we were able to guarantee that this cleanup happened _before_ starting a new pod.

e.g 
- Kubelet decides to start a pod
- Kubelet checks that no dead containers correspond with that pod
- Kubelet tears down any containers it finds
- Kubelet starts the pod.

Not familiar enough with that bit of code to actually know if that's reasonable.

For Calico in particular, we store various bits of information (IPAM allocation, labels, etc) related to a particular pod essentially keyed off of the pod_namespace+pod_name.  This has allowed the CNI plugin to detect when a pod has been restarted vs when it's a new pod.  It's also useful for people using the Calico APIs to see data represented in terms of pod names rather than container IDs.

> Also, does the existing Calico workaround handle the case where the pod does not get rescheduled on the same node, but the net container is still dead with its IP allocation intact?

Nope, it does not.  It only works if the node comes back online before k8s decides to reschedule, which is why something like this PR will still be really useful.  If we can do something like I suggested above that would be really great.

freehan replied:
GC and syncPod are separate in kubelet. From kubelet's point of view, GC dead infra container and creating new infra container are irrelevant.

dcbw replied:
@caseydavenport @freehan I was going to do a follow-on to this for the docker runtime that ensured that a previous infra container *was* cleaned up before starting a new one.  Currently the docker runtime only cleans up running infra containers:

```
func (dm *DockerManager) SyncPod(...) {
	<snip>
	if containerChanges.StartInfraContainer || (len(containerChanges.ContainersToKeep) == 0 && len(containerChanges.ContainersToStart) == 0) {
		if len(containerChanges.ContainersToKeep) == 0 && len(containerChanges.ContainersToStart) == 0 {
			glog.V(4).Infof("Killing Infra Container for %q because all other containers are dead.", format.Pod(pod))
		} else {
			glog.V(4).Infof("Killing Infra Container for %q, will start new one", format.Pod(pod))
		}

		killResult := dm.killPodWithSyncResult(pod, kubecontainer.ConvertPodStatusToRunningPod(dm.Type(), podStatus), nil)
```

unfortunately that kubecontainer.ConvertPodStatusToRunningPod() call only converts running containers, so the dead infra is left to GC and not cleaned up.  Perhaps if we also solve this, it will fix the possible conflict with plugin workarounds by ensuring that the dead infra always gets cleaned up before a new one is started, as you suggest?

feiskyer replied:
> unfortunately that kubecontainer.ConvertPodStatusToRunningPod() call only converts running containers, so the dead infra is left to GC and not cleaned up. Perhaps if we also solve this, it will fix the possible conflict with plugin workarounds by ensuring that the dead infra always gets cleaned up before a new one is started, as you suggest?

SyncPod only kills unwanted running containers, and all dead containers are left to GC. I'm concerned of introducing much sync latency if killing dead containers are added into syncPod.

caseydavenport replied:
> and all dead containers are left to GC

It wouldn't have to cleanup _all_ dead containers.. just any dead container that is relevant to the pod being synced (i.e a previous "pause" container).  Do we think that cleaning up a single container will introduce too much latency?  Also I would think this shouldn't introduce any latency in mainline "I launched a new pod" type scenarios, this should only happen when a running pod is killed by something external (e.g docker / the node restarting)

dcbw replied:
@caseydavenport @feiskyer I agree with Casey, killing a dead infra container during sync is an exceptional event, not a normal one.  So I think that would be OK.

dcbw committed:
dockertools: tear down dead infra containers in SyncPod()

Dead infra containers may still have network resources allocated to
them and may not be GC-ed for a long time.  But allowing SyncPod()
to restart an infra container before the old one is destroyed
prevents network plugins from carrying the old network details
(eg IPAM) over to the new infra container.

dcbw commented:
@caseydavenport @feiskyer @freehan @thockin I've implemented the dead infra container teardown for SyncPod() now.  PTAL thanks!

feiskyer commented:
@k8s-bot kops aws e2e test this

feiskyer commented:
LGTM

resouer commented:
@dcbw could you please squash your commits and fix rebase? Then leave to assignees to make final decision.

dcbw commented:
@freehan @yujuhong @caseydavenport rebased, PTAL thanks!

dcbw commented:
@dchen1107 the merge bot wants your approval...

caseydavenport commented:
@dcbw finally got another chance to review - this LGTM, thanks!

dcbw commented:
@freehan all rebased and ready for re-review, thanks!

freehan commented:
LGTM Thanks!

freehan commented:
Added a priority since rebased more than 2 times

dcbw commented:
@freehan rebased!

freehan commented:
@k8s-bot cross build this
@k8s-bot cvm gce e2e test this
@k8s-bot gce etcd3 e2e test this

yujuhong commented:
/approve

saiwl commented:
When will this feature be released? We really need it.

yujuhong commented:
@saiwl this is already included in v1.6.

saiwl commented:
Is this feature turned on by default? I restarted the docker daemon, but the ip wasn't released. @yujuhong

yujuhong commented:
> Is this feature turned on by default? I restarted the docker daemon, but the ip wasn't released. @yujuhong

This is turned on by default (or simply put it, there is no way to disable this). 

If you want to report a bug, please open a new issue with more information (e.g., environment, network setup, steps to reproduce, etc), thanks.
